src/go/k8s/internal/controller/redpanda/cluster_controller.go

```diff
@@ -361,7 +374,7 @@ func (r *ClusterReconciler) handlePodFinalizer(
 					untainted = false
 				}
 			}
-			if untainted {
+			if untainted && !decommissionOnDelete {
```

Comment:
I'm thinking we should probably make this safer based on the description/use-case of your PR and the scenarios at hand.

Say we've set the annotation on the cluster (ie for an upgrade that takes hours) and let's consider the following scenarios:
1. During the process, a human operator sets a configuration (requiring a restart) using rpk on the cluster. That would trigger the operator to do a whole-cluster restart. Would that delete and recreate all of the PVCs sequentially?
2. Say due to an update/patch error or urgency of incidents the annotation is left on the cluster. Would that trigger PVC recreation on every pod delete (eg restarts/upgrades etc)?

Based on the usecases on the PR (upgrades) and the fact that the main process used there is node draining would it make sense to keep and extend the current logic and just respect the `NoSchedule` taint if the annotation is set?
ie adding a new case to the for-loop directly above
```go
if decomissionOnDelete && taint.Effect == corev1.TaintEffectNoSchedule {
    untainted = false
}
```

That would give us more confidence we don't decomission nodes that aren't going away and have to transfer data "by mistake".